contracts/utils/RetireToucanCarbon.sol

```diff
@@ -0,0 +1,1625 @@
+        // the target address contains contract code and also asserts for success in the low-level call.
+
+        bytes memory returndata = address(token).functionCall(
+            data,
+            "SafeERC20: low-level call failed"
+        );
+        if (returndata.length > 0) {
+            // Return data is optional
+            require(
+                abi.decode(returndata, (bool)),
+                "SafeERC20: ERC20 operation did not succeed"
+            );
+        }
+    }
+}
+
+// File: contracts/utils/SafeERC20.sol
+
+// File: contracts/RetireToucanCarbon.sol
+
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+contract RetireToucanCarbon is Ownable {
+    using SafeERC20 for IERC20;
+
+    constructor(
+        address _KLIMA,
+        address _sKLIMA,
+        address _wsKLIMA,
+        address _USDC,
+        address _staking,
+        address _stakingHelper,
+        address _DAO,
+        address _KlimaRetirementStorage
+    ) {
+        // Null check for constructors
+        require(_KLIMA != address(0));
+        require(_sKLIMA != address(0));
+        require(_wsKLIMA != address(0));
+        require(_USDC != address(0));
+        require(_staking != address(0));
+        require(_stakingHelper != address(0));
+        require(_DAO != address(0));
+        require(_KlimaRetirementStorage != address(0));
+
+        KLIMA = _KLIMA;
+        sKLIMA = _sKLIMA;
+        wsKLIMA = _wsKLIMA;
+        USDC = _USDC;
+        staking = _staking;
+        stakingHelper = _stakingHelper;
+        DAO = _DAO;
+        KlimaRetirementStorage = _KlimaRetirementStorage;
+    }
+
+    address public KLIMA;
+    address public sKLIMA;
+    address public wsKLIMA;
+    address public immutable USDC;
+    address public staking;
+    address public stakingHelper;
+    address public immutable DAO;
+    address public KlimaRetirementStorage;
+
+    // Event setup
+    event ToucanRetired(
+        address indexed retiringAddress,
+        address indexed beneficiaryAddress,
+        string beneficiaryString,
+        string retirementMessage,
+        address indexed carbonPool,
+        address carbonToken,
+        uint256 retiredAmount
+    );
+    event PoolAdded(
+        address indexed carbonPool,
+        address indexed poolRouter,
+        address indexed bondDepository
+    );
+    event PoolRouterChanged(
+        address indexed carbonPool,
+        address indexed oldRouter,
+        address indexed newRouter
+    );
+    event PoolBondChanged(
+        address indexed carbonPool,
+        address indexed oldBondDepository,
+        address indexed newBondDepository
+    );
+    event FeeUpdated(uint256 oldFee, uint256 newFee);
+
+    address[] public toucanPools;
+    mapping(address => bool) public isToucanPool;
+    mapping(address => address) public poolRouter;
+    mapping(address => address) public bondDepository;
```

Comment:
This array and three mappings can be combined:
```
    struct ToucanPool{
        address poolRouter;
        address bondDepository;
    }
    mapping (address => ToucanPool) public toucanPools;
```

Also you will safe tons of gas not iterating through `toucanPools` checking for duplicates via `listContains()`